kqueen/auth/ldap.py

```diff
@@ -61,30 +92,42 @@ def verify(self, user, password):
             None: Invalid credentials, Authentication failure.
             User: authenticated user.
         """
+        matched_dn = self._get_matched_dn(user.username)
+
+        for dn in matched_dn:
+            if self._bind(dn, password):
+                full_dn = dn
+
+        if full_dn:
+            logger.debug('Valid full-DN found: {}. It will be stored in user metadata'.format(full_dn))
+            return user, None
+
+        return None, "Verification failed"
 
-        dn = self._email_to_dn(user.username)
+    def _bind(self, dn, password):
 
         try:
+            self.connection = ldap.initialize(self.uri)
             bind = self.connection.simple_bind_s(dn, password)
 
             if bind:
-                return user, None
+                return True
         except ldap.INVALID_CREDENTIALS:
             logger.exception("Invalid LDAP credentials for {}".format(dn))
 
-            return None, "Invalid LDAP credentials"
+            return False, "Invalid LDAP credentials"
 
         except ldap.INVALID_DN_SYNTAX:
             logger.exception("Invalid DN syntax in configuration: {}".format(dn))
 
-            return None, "Invalid DN syntax"
+            return False, "Invalid DN syntax"
 
         except ldap.LDAPError:
             logger.exception("Failed to bind LDAP server")
 
-            return None, "LDAP auth failed, check log for error"
+            return False, "LDAP auth failed, check log for error"
 
```

Comment:
You might want to handle generic `Exception` in the end, because you might lose network connection for example

Reply:
seems your point is not clear for me , pls clarify info

Reply:
You handle various LDAP errors in the try/except block, but if you would for example lose network connection and you are unable to reach the LDAP server, then try block fails, but none of these except blocks will catch the `unreachable` exception. So I would add `except Exception as e` in the end and create message like `msg = 'Unknown error occurred: {}'.format(repr(e))` or something. So I handle even the exceptions you don't necessarily expect to occur